app/dashboard/src/layouts/AssetsTable.tsx

```diff
@@ -1417,68 +1369,43 @@ export default function AssetsTable(props: AssetsTableProps) {
-  }, [])
+  const deleteAsset = useEventCallback((assetId: backendModule.AssetId) => {
+    const asset = nodeMapRef.current.get(assetId)?.item
+
+    if (asset) {
+      const listDirectoryQuery = queryClient.getQueryCache().find<ListDirectoryQueryDataType>({
+        queryKey: ['listDirectory', backend.type, asset.parentId],
+        exact: false,
+      })
+
+      if (listDirectoryQuery?.state.data) {
+        listDirectoryQuery.setData({
+          ...listDirectoryQuery.state.data,
+          children: listDirectoryQuery.state.data.children.filter((child) => child.id !== assetId),
+        })
+      }
+    }
+  })
 
   /** All items must have the same type. */
-  const insertAssets = React.useCallback(
-    (
-      assets: readonly backendModule.AnyAsset[],
-      parentKey: backendModule.DirectoryId | null,
-      parentId: backendModule.DirectoryId | null,
-      getInitialAssetEvents: (id: backendModule.AssetId) => readonly assetEvent.AssetEvent[] | null,
-    ) => {
-      const actualParentKey = parentKey ?? rootDirectoryId
+  const insertAssets = useEventCallback(
+    (assets: readonly backendModule.AnyAsset[], parentId: backendModule.DirectoryId | null) => {
       const actualParentId = parentId ?? rootDirectoryId
-      setAssetTree((oldAssetTree) =>
-        oldAssetTree.map((item) =>
-          item.key !== actualParentKey ?
-            item
-          : insertAssetTreeNodeChildren(
-              item,
-              assets,
-              actualParentKey,
-              actualParentId,
-              getInitialAssetEvents,
-            ),
-        ),
-      )
-    },
-    [rootDirectoryId],
-  )
 
-  const insertArbitraryAssets = React.useCallback(
-    (
-      assets: backendModule.AnyAsset[],
-      parentKey: backendModule.DirectoryId | null,
-      parentId: backendModule.DirectoryId | null,
-      getKey: ((asset: backendModule.AnyAsset) => backendModule.AssetId) | null = null,
-      getInitialAssetEvents: (
-        id: backendModule.AssetId,
-      ) => readonly assetEvent.AssetEvent[] | null = () => null,
-    ) => {
-      const actualParentKey = parentKey ?? rootDirectoryId
-      const actualParentId = parentId ?? rootDirectoryId
-      setAssetTree((oldAssetTree) => {
-        return oldAssetTree.map((item) =>
-          item.key !== actualParentKey ?
-            item
-          : insertArbitraryAssetTreeNodeChildren(
-              item,
-              assets,
-              actualParentKey,
-              actualParentId,
-              getKey,
-              getInitialAssetEvents,
-            ),
-        )
+      const listDirectoryQuery = queryClient.getQueryCache().find<ListDirectoryQueryDataType>({
+        queryKey: ['listDirectory', backend.type, actualParentId],
+        exact: false,
       })
+
+      if (listDirectoryQuery?.state.data) {
+        listDirectoryQuery.setData({
+          ...listDirectoryQuery.state.data,
+          children: [...listDirectoryQuery.state.data.children, ...assets],
```

Comment:
optional, re: ordering -
see https://github.com/enso-org/enso/pull/10775/files?w=1#diff-0e343617d95ffe3a6da1c565633d686bb84ec06614b5bd8b4833c56d78d1b198L225 for how it was previously done